### ISSUE TITLE
Compatibility with ES 2.x 

### ISSUE THREAD

**Issue by tomcgn** [jprante/elasticsearch-oai#11]
@xbib, is there a roadmap for the integration with future  elasticsearch releases? What would be a good  starting point, if one wanted to make the latest version compatible to ES 2.0 / 2.1 ?

**Comment by jprante** [comment]
Thanks for the request. I have just released OAI 2.1.0.0 for Elasticsearch 2.1

**Comment by tomcgn** [comment]
Hi Jörg, great, thanks for the quick update! 
`mvn install` creates a standalone version and a plugin. Is the plugin supposed to be installed on the elasticsearch instance that will later be fed with metadata? 
I can run the arxiv example, but no data is received on the elasticsearch side. 
Is there a simle way to run a harvest session from eclipse for testing?

**Comment by jprante** [comment]
The plugin mode does not work, and is just in the code base as a placeholder for Elasticsearch >= 2.2 when a task API will be added so OAI harvest jobs might be possible to execute within a node, similar to obsoleted river API. Also I intend to add OAI server mode.
Arxiv does not work for me because they are very aggressive about 20 seconds pause. They always respond with "Retry-After" - I'm clueless. Other services should work better, like DOAJ. XML to ES JSON mapping could be done better.
I don't use Eclipse, sorry. Also sorry about missing junit test code. It should be possible to set up a small java program which executes the `OAIFeeder` if you mean that.

**Comment by tomcgn** [comment]
Ok, nice that you are heading the task API! 
I used a different OAI endpoint, and even with arxiv I can see that there are responses coming in via PMH. The issues seems related to the actual feeding into Elasticsearch. 
Due to the recommendation to have only one Elasticsearch server per machine I started to use docker images. I will install elasticsearch-oai on a ubuntu image and thus be fine with the shell skripts. 

The overall task I'd like to solve is to incrementally harvest several oai repositories, that provide metadata in [the ddi format](http://oai.esds.ac.uk/oai/provider?verb=GetRecord&metadataPrefix=ddi&identifier=6512). 

Before I saw your project, I tried to ingest dc into elasticsearch with a small spring application. 
That went all fine, fast and except some date format parsing without major elasticsearch schema editing. 

It got difficult when I wanted to harvest ddi metadata  and deal with the [different versions and specifics of the ddi format](http://www.ddialliance.org/explore-documentation) : How to handle xml attributes and nested elements? How to translate those into an appropriate elasticsearch mapping [so that different languages for controlled vocabularies remain useful and do not get flattened](https://www.elastic.co/guide/en/elasticsearch/reference/2.0/nested.html), for example? 
The same applies to the same xml element occuring as child element of different parent elements, such as `StartDate`. 

Does your solution adress those issues?

**Comment by jprante** [comment]
Thanks for the info about DDI format. I will study it to learn the challenges.

For transforming XML into ES, I have some methods in my xbib project, which are more or less promising.

- simple XML i.e. only element names, no namespaces, no arrays, no attributes: this can be translated straight forward to a key/value stream and then to ES JSON

- XML with peculiarities i.e. key/value mapping not straight: adding a "mapper" class that can do stateful transformation (renaming elements, fixing element data) so ES JSON can be generated

- full fledged XML with namespaces, (nested) arrays, attributes, etc.: because there is no simple key/value stream, I can try to parse this XML to RDF. The generation of RDF should be constrained by a domain model to mark arrays, maintain order of elements etc. The RDF graph can be flattened to a key/value stream and indexed as ES JSON.  If the original document must be reconstructed later after ES retrieval, there is extra work to do. 

Parsing XML to RDF is not perfect yet, I have challenges with attribute names and nesting, there is still much work to do. It's frustrating and time consuming to say the least.

**Comment by tomcgn** [comment]
I agree, the last of the three use cases is a bit special. I followed the mapping approach in my prototype, too. 

Ideally one could create an elasticsearch schema either from the .xsd or java beans and simply feed xml to elasticsearch. 

I had java beans generated for ddi 3.1 with xjc ([there are packages for 3.1](http://www.ddialliance.org/tool/xml-beans-for-ddi-30)[and 1.2](http://www.ddialliance.org/tool/xml-beans-for-ddi-12x), but not for the current v3.1). 

So the processing would contain of simply first unmarshal the metadata to beans, second inject into elasticsearch. 
One way to unmarshal from xml is the `com.fasterxml.jackson.dataformat.xml.XmlMapper` which has one drawback: It populates only the fist element, lets say [DDIInstance](http://www.ddialliance.org/system/files/ICPSR2079variables.xml) with its own attributes, but the nested elements are not fetched.

So to come back to your third point:
We do not need to reconstruct the original document. But yes, the parent-child and element-[attributes relation needs to be preserved](https://www.elastic.co/guide/en/elasticsearch/reference/2.0/nested.html). The domain model that you mention for the RDF generation, could this be derived from the JAXB/xjc created bean classes resp. the .xsd files? 

As elasticsearch _does_ provide a way to address nested objects, there should be a way to code a mapping / mapper generator.

**Comment by jprante** [comment]
It's very complex to generate RDF from existing XML and the result is most always frustrating.
I wonder if mapping the data to http://www.ddialliance.org/Specification/RDF/Discovery is more promising, for instance, because of the statement " Existing DDI XML instances can be transformed into this RDF format and therefore exposed in the Web of Linked Data. " - but I don't know how to do so. 
For the general case, I don't know any solution to the domain mapping problem.

**Comment by ynux** [comment]
About arXiv: It seems to work better when you restrict to a set, i.e. add "&set=physics:hep-ex"
